### PR TITLE
Fix the way to get username and password.

### PR DESCRIPTION
**Pull request by nicelulu**
When I executed sql: "ALTER TABLE search. logs DETACH PARTITION 20181103", I received this exception: "DB:: Exception: Password required for user default".

When sending requests to leader, member variable context does not get the correct username password. We should use context in the dropPartition method parameter or get the user name and password of leader from the cluster configuration. Truncate also has similar problems.

Here, I used the way to get the username and password of leader from the cluster configuration to fix this.

### CODE REVIEW

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -3885,6 +3886,30 @@ void StorageReplicatedMergeTree::sendRequestToLeaderReplica(const ASTPtr & query
     return;
 }
 
+const Cluster::Address & StorageReplicatedMergeTree::findClusterAddress(ReplicatedMergeTreeAddress & leader_address) const
+{
+    const auto & clusters = context.getClusters();
+    const auto & clusterPtrs = clusters.getContainer();
+
+    for(auto iter = clusterPtrs.begin(); iter != clusterPtrs.end(); ++iter)
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.cpp]: Style.

File: dbms/src/Storages/StorageReplicatedMergeTree.h [hunk 2]
```diff
@@ -512,6 +513,8 @@ class StorageReplicatedMergeTree : public ext::shared_ptr_helper<StorageReplicat
     bool dropPartsInPartition(zkutil::ZooKeeper & zookeeper, String & partition_id,
         StorageReplicatedMergeTree::LogEntry & entry, bool detach);
 
+    const Cluster::Address & findClusterAddress(ReplicatedMergeTreeAddress & leader_address) const;
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.h]: const?

Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.h]: Missing comment.

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -3870,13 +3871,13 @@ void StorageReplicatedMergeTree::sendRequestToLeaderReplica(const ASTPtr & query
         throw Exception("Can't proxy this query. Unsupported query type", ErrorCodes::NOT_IMPLEMENTED);
 
     /// Query send with current user credentials
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.cpp]: Obsolete comment.

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Storages/StorageReplicatedMergeTree.cpp]: You can use range based loop.

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp [hunk 4]
```diff
@@ -3885,6 +3886,30 @@ void StorageReplicatedMergeTree::sendRequestToLeaderReplica(const ASTPtr & query
     return;
 }
 
+const Cluster::Address & StorageReplicatedMergeTree::findClusterAddress(ReplicatedMergeTreeAddress & leader_address) const
+{
+    const auto & clusters = context.getClusters();
+    const auto & clusterPtrs = clusters.getContainer();
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.cpp]: Style.

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp [hunk 5]
```diff
@@ -3885,6 +3886,30 @@ void StorageReplicatedMergeTree::sendRequestToLeaderReplica(const ASTPtr & query
...
+                const Cluster::Address & address = shards[shard_num][replica_num];
+
+                if (address.host_name == leader_address.host && address.port == leader_address.queries_port)
+                {
+                    return address;
+                }
+            }
+        }
+    }
+    throw Exception("Not found host " + leader_address.host, ErrorCodes::INCONSISTENT_CLUSTER_DEFINITION);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.cpp]: The error message is not descriptive enough.

**Comment by alexey-milovidov** [comment]
Missing integration test.

**Comment by zhang2014** [comment]
@alexey-milovidov Maybe use `context.getInterserverCredentials()` is better? But that may require some refactor

**Comment by zhang2014** [comment]
How to choose username and password :
``` xml
<remote_servers>
    <database_a_cluster>
        <shard>
            <replica>
                <host>192.168.0.2</host>
                <user>username_a</user>
                <password>password_a</password>
            </replica>
        </shard>
    </database_a_cluster>
    <database_b_cluster>
        <shard>
            <replica>
                <host>192.168.0.2</host>
                <user>username_b</user>
                <password>password_b</password>
            </replica>
        </shard>
    </database_b_cluster>
</remote_servers>
```
 
Maybe we can refactor `InterserverCredentials` to support queries ?

**Comment by nicelulu** [comment]
I am trying to complete the integration test.I executed: "pytest -k test_replace_partition".
The following error occurred.


>/python2.7/subprocess.py:542: CalledProcessError
--------------------------------------------------------------------------------------------------------------------------- Captured stderr setup ----------------------------------------------------------------------------------------------------------------------------
Removing network clickhouse_default
Network clickhouse_default not found.
.IOError: [Errno 2] No such file or directory: '/data0/dev/ClickHouse/dbms/tests/integration/test_replace_partition/_instances/node1/docker_compose.yml'
____________________________________________________________________________________________________________________ ERROR at setup of test_drop_failover ____________________________________________________________________________________________________________________

    @pytest.fixture(scope="module")
    def drop_failover():
        try:
            cluster.start()
    
            _fill_nodes([node3, node4], 2)
    
            yield cluster
    
        finally:
    >       cluster.shutdown()

CentOS Linux release 7.5.1804
Docker version 18.09.0, build 4d60db4
docker-compose version 1.23.1, build b02f130
This is pytest version 4.0.0

Excuse me, where is this problem? Thank you.

**Comment by nicelulu** [comment]
@alesapin 
Hello, thank you.
I have some such errors now.
"NotFound: 404 Client Error: Not Found ("No such container: roottestreplacepartition_node5_1")"
docker_client.containers.get(docker_id) did not get the container. I don't know why docker_id is roottestreplacepartition_node5_1.
The docker_id is expected to be like this 45e6d2de7c54.
There is another problem: "E Exception: Instance `node1' failed to start. Container status: exited, logs: Include not found: clickhouse_remote_servers
E Include not found: clickhouse_compression
E Logging trace to /var/log/clickhouse-server/clickhouse-server.log
E Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
E Include not found: networks
E Include not found: networks
E Include not found: networks
"There are some configurations for the service in the container that are not found. I don't know how the configuration file under /dbms/tests/integration/test_replace_partition/configs/ is combined with /etc/clickhouse-server/config.xml.
There is another problem. The initial location of the pytest program seems to be @pytest.fixture, but for example some ClickHouseCluster initialization is before it. I don't know when it was called. I need some help, thanks.

**Comment by nicelulu** [comment]
@alesapin 
Hello,thank you.
This is with -ss option and provide full output.
```
[root@server50-ke integration]# pytest -ss -k test_replicated_drop_partition
============================================================================================================================ test session starts =============================================================================================================================
platform linux2 -- Python 2.7.5, pytest-4.0.0, py-1.7.0, pluggy-0.8.0
rootdir: /data0/dev/ClickHouse/dbms/tests/integration, inifile: pytest.ini
collected 91 items / 90 deselected                                                                                                                                                                                                                                           

test_replicated_drop_partition/test.py Removing network clickhouse_default
WARNING: Network clickhouse_default not found.
Removing instances dir /data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition/_instances
suss
ERROR: .IOError: [Errno 2] No such file or directory: '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition/_instances/node1/docker_compose.yml'
E

=================================================================================================================================== ERRORS ===================================================================================================================================
_____________________________________________________________________________________________________________________ ERROR at setup of test_normal_work _____________________________________________________________________________________________________________________

    @pytest.fixture(scope="module")
    def normal_work():
        try:
    	cluster.start()
    
            _fill_nodes([node1, node2],[default,default],[pTe5Tb0s,dtnDvTr9],1)
    	_fill_nodes([node2, node3],[zhang,default],[azAUGBFl,ROgXGTDq],2)
    
            yield cluster
    
        finally:
    	print('suss')
>           cluster.shutdown()

test_replicated_drop_partition/test.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers/cluster.py:276: in shutdown
    subprocess_check_call(self.base_cmd + ['kill'])
helpers/cluster.py:37: in subprocess_check_call
    subprocess.check_call(args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

popenargs = (['docker-compose', '--project-directory', '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition', '--project-name', 'roottestreplicateddroppartition', '--file', ...],), kwargs = {}, retcode = 1
cmd = ['docker-compose', '--project-directory', '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition', '--project-name', 'roottestreplicateddroppartition', '--file', ...]

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.
    
        The arguments are the same as for the Popen constructor.  Example:
    
        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           CalledProcessError: Command '['docker-compose', '--project-directory', '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition', '--project-name', 'roottestreplicateddroppartition', '--file', '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition/_instances/node1/docker_compose.yml', '--file', '/data0/dev/ClickHouse/dbms/tests/integration/helpers/docker_compose_zookeeper.yml', '--file', '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition/_instances/node2/docker_compose.yml', '--file', '/data0/dev/ClickHouse/dbms/tests/integration/test_replicated_drop_partition/_instances/node3/docker_compose.yml', 'kill']' returned non-zero exit status 1

/usr/lib64/python2.7/subprocess.py:542: CalledProcessError
=================================================================================================================== 90 deselected, 1 error in 1.92 seconds ===================================================================================================================
```

**Comment by alesapin** [comment]
I think you have some misconfiguration in your system, but now you can use [runner](https://github.com/yandex/ClickHouse/tree/master/dbms/tests/integration#running-with-runner-script) script in dbms/test/integration for integration tests. It doesn't require any dependencies and configuration -- only docker. Try it out.

**Comment by nicelulu** [comment]
@alesapin 
Thank you.
I still have some errors.
```
[root@server50-ke integration]# ./runner --binary /data0/dev/ClickHouse/dbms/programs/clickhouse --configs-dir /data0/dev/ClickHouse/dbms/programs/server/ 'test_odbc_interaction -ss'
Start tests
ImportError while loading conftest '/ClickHouse/dbms/tests/integration/conftest.py'.
ImportMismatchError: ('conftest', '/data0/dev/ClickHouse/dbms/tests/integration/conftest.py', local('/ClickHouse/dbms/tests/integration/conftest.py'))
Traceback (most recent call last):
  File "./runner", line 55, in <module>
    subprocess.check_call(cmd, shell=True)
  File "/usr/lib64/python2.7/subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'docker run --net=host --privileged --volume=/data0/dev/ClickHouse/dbms/programs/clickhouse:/clickhouse         --volume=/data0/dev/ClickHouse/dbms/programs/server/:/clickhouse-config --volume=/data0/dev/ClickHouse:/ClickHouse -e PYTEST_OPTS='test_odbc_interaction -ss' yandex/clickhouse-integration-tests-runner' returned non-zero exit status 4
```

**Comment by nicelulu** [comment]
@alesapin 
Thank you.
When I executed the runner, I encountered the following error.
```
[root@server50-ke integration]# ./runner --binary /data0/dev/ClickHouse/build/dbms/programs/clickhouse --configs-dir /usr/local/etc/clickhouse-server/ 'test_replace_partition -ss'
Start tests
============================= test session starts ==============================
platform linux2 -- Python 2.7.15rc1, pytest-4.0.0, py-1.7.0, pluggy-0.8.0
rootdir: /ClickHouse/dbms/tests/integration, inifile: pytest.ini
collected 3 items

test_replace_partition/test.py Removing network clickhouse_default
Network clickhouse_default not found.
Creating network "roottestreplacepartition_default" with the default driver
Pulling zoo1 (zookeeper:)...
latest: Pulling from library/zookeeper
Digest: sha256:3f03c6f5a91e0f638f3d6a755b2d32c06583766031353be87e2d633fa3006c23
Status: Downloaded newer image for zookeeper:latest
Creating roottestreplacepartition_zoo1_1 ... done
Creating roottestreplacepartition_zoo3_1 ... done
Creating roottestreplacepartition_zoo2_1 ... done
Pulling node1 (yandex/clickhouse-integration-test:)...
latest: Pulling from yandex/clickhouse-integration-test
Digest: sha256:d77fe4c0baaa7f0f07399d28a451ae02e70c685d3068ea4bd22782066cf33107
Status: Downloaded newer image for yandex/clickhouse-integration-test:latest
Recreating roottestreplacepartition_zoo3_1 ... done
Recreating roottestreplacepartition_zoo2_1 ... done
Recreating roottestreplacepartition_zoo1_1 ... done
Creating roottestreplacepartition_node4_1  ... done
Creating roottestreplacepartition_node3_1  ... done
Creating roottestreplacepartition_node6_1  ... done
Creating roottestreplacepartition_node1_1  ... done
Creating roottestreplacepartition_node2_1  ... done
Creating roottestreplacepartition_node5_1  ... done
Killing roottestreplacepartition_zoo2_1 ... done
Killing roottestreplacepartition_zoo3_1 ... done
Killing roottestreplacepartition_zoo1_1 ... done
Removing roottestreplacepartition_node5_1 ... done
Removing roottestreplacepartition_node2_1 ... done
Removing roottestreplacepartition_node1_1 ... done
Removing roottestreplacepartition_node3_1 ... done
Removing roottestreplacepartition_node6_1 ... done
Removing roottestreplacepartition_node4_1 ... done
Removing roottestreplacepartition_zoo2_1  ... done
Removing roottestreplacepartition_zoo3_1  ... done
Removing roottestreplacepartition_zoo1_1  ... done
Removing network roottestreplacepartition_default
All instances of ZooKeeper started
ERemoving network clickhouse_default
Network clickhouse_default not found.
Creating network "roottestreplacepartition_default" with the default driver
Creating roottestreplacepartition_zoo2_1 ... done
Creating roottestreplacepartition_zoo3_1 ... done
Creating roottestreplacepartition_zoo1_1 ... done
Recreating roottestreplacepartition_zoo2_1 ... done
Recreating roottestreplacepartition_zoo3_1 ... done
Recreating roottestreplacepartition_zoo1_1 ... done
Creating roottestreplacepartition_node2_1  ... done
Creating roottestreplacepartition_node3_1  ... done
Creating roottestreplacepartition_node5_1  ... done
Creating roottestreplacepartition_node4_1  ... done
Creating roottestreplacepartition_node6_1  ... done
Creating roottestreplacepartition_node1_1  ... done
Killing roottestreplacepartition_zoo1_1 ... done
Killing roottestreplacepartition_zoo3_1 ... done
Killing roottestreplacepartition_zoo2_1 ... done
Removing roottestreplacepartition_node5_1 ... done
Removing roottestreplacepartition_node1_1 ... done
Removing roottestreplacepartition_node4_1 ... done
Removing roottestreplacepartition_node3_1 ... done
Removing roottestreplacepartition_node2_1 ... done
Removing roottestreplacepartition_node6_1 ... done
Removing roottestreplacepartition_zoo1_1  ... done
Removing roottestreplacepartition_zoo3_1  ... done
Removing roottestreplacepartition_zoo2_1  ... done
Removing network roottestreplacepartition_default
Removing instances dir /ClickHouse/dbms/tests/integration/test_replace_partition/_instances
All instances of ZooKeeper started
ERemoving network clickhouse_default
Network clickhouse_default not found.
Creating network "roottestreplacepartition_default" with the default driver
Creating roottestreplacepartition_zoo2_1 ... done
Creating roottestreplacepartition_zoo1_1 ... done
Creating roottestreplacepartition_zoo3_1 ... done
Recreating roottestreplacepartition_zoo3_1 ... done
Recreating roottestreplacepartition_zoo1_1 ... done
Recreating roottestreplacepartition_zoo2_1 ... done
Creating roottestreplacepartition_node4_1  ... done
Creating roottestreplacepartition_node2_1  ... done
Creating roottestreplacepartition_node1_1  ... done
Creating roottestreplacepartition_node6_1  ... done
Creating roottestreplacepartition_node5_1  ... done
Creating roottestreplacepartition_node3_1  ... done
Killing roottestreplacepartition_zoo2_1 ... done
Killing roottestreplacepartition_zoo3_1 ... done
Killing roottestreplacepartition_zoo1_1 ... done
Removing roottestreplacepartition_node1_1 ... done
Removing roottestreplacepartition_node4_1 ... done
Removing roottestreplacepartition_node3_1 ... done
Removing roottestreplacepartition_node5_1 ... done
Removing roottestreplacepartition_node6_1 ... done
Removing roottestreplacepartition_node2_1 ... done
Removing roottestreplacepartition_zoo2_1  ... done
Removing roottestreplacepartition_zoo3_1  ... done
Removing roottestreplacepartition_zoo1_1  ... done
Removing network roottestreplacepartition_default
Removing instances dir /ClickHouse/dbms/tests/integration/test_replace_partition/_instances
All instances of ZooKeeper started
E

==================================== ERRORS ====================================
______________________ ERROR at setup of test_normal_work ______________________

    @pytest.fixture(scope="module")
    def normal_work():
        try:
>           cluster.start()

test_replace_partition/test.py:35: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers/cluster.py:257: in start
    instance.wait_for_start(start_deadline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <helpers.cluster.ClickHouseInstance instance at 0x7fe06e71ef38>
deadline = 1543320505.437261, timeout = None

    def wait_for_start(self, deadline=None, timeout=None):
        start_time = time.time()
    
        if timeout is not None:
            deadline = start_time + timeout
    
        while True:
            handle = self.get_docker_handle()
            status = handle.status;
            if status == 'exited':
>               raise Exception("Instance `{}' failed to start. Container status: {}, logs: {}".format(self.name, status, handle.logs()))
E               Exception: Instance `node1' failed to start. Container status: exited, logs: clickhouse: error while loading shared libraries: libreadline.so.6: cannot open shared object file: No such file or directory

helpers/cluster.py:425: Exception
------------------------------ Captured log setup ------------------------------
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection reset by peer
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection reset by peer
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection reset by peer
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
_____________________ ERROR at setup of test_drop_failover _____________________

    @pytest.fixture(scope="module")
    def drop_failover():
        try:
>           cluster.start()

test_replace_partition/test.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers/cluster.py:257: in start
    instance.wait_for_start(start_deadline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <helpers.cluster.ClickHouseInstance instance at 0x7fe06e71ef38>
deadline = 1543320512.590594, timeout = None

    def wait_for_start(self, deadline=None, timeout=None):
        start_time = time.time()
    
        if timeout is not None:
            deadline = start_time + timeout
    
        while True:
            handle = self.get_docker_handle()
            status = handle.status;
            if status == 'exited':
>               raise Exception("Instance `{}' failed to start. Container status: {}, logs: {}".format(self.name, status, handle.logs()))
E               Exception: Instance `node1' failed to start. Container status: exited, logs: clickhouse: error while loading shared libraries: libreadline.so.6: cannot open shared object file: No such file or directory

helpers/cluster.py:425: Exception
------------------------------ Captured log setup ------------------------------
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
____________ ERROR at setup of test_replace_after_replace_failover _____________

    @pytest.fixture(scope="module")
    def replace_after_replace_failover():
        try:
>           cluster.start()

test_replace_partition/test.py:109: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers/cluster.py:257: in start
    instance.wait_for_start(start_deadline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <helpers.cluster.ClickHouseInstance instance at 0x7fe06e71ef38>
deadline = 1543320518.596473, timeout = None

    def wait_for_start(self, deadline=None, timeout=None):
        start_time = time.time()
    
        if timeout is not None:
            deadline = start_time + timeout
    
        while True:
            handle = self.get_docker_handle()
            status = handle.status;
            if status == 'exited':
>               raise Exception("Instance `{}' failed to start. Container status: {}, logs: {}".format(self.name, status, handle.logs()))
E               Exception: Instance `node1' failed to start. Container status: exited, logs: clickhouse: error while loading shared libraries: libreadline.so.6: cannot open shared object file: No such file or directory

helpers/cluster.py:425: Exception
------------------------------ Captured log setup ------------------------------
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              612 WARNING  Transition to CONNECTING
connection.py              608 WARNING  Connection dropped: socket connection error: None
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: Connection refused
connection.py              608 WARNING  Connection dropped: socket connection error: None
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              608 WARNING  Connection dropped: socket connection broken
connection.py              608 WARNING  Connection dropped: socket connection error: None
=============================== warnings summary ===============================
/usr/local/lib/python2.7/dist-packages/psycopg2/__init__.py:144
  /usr/local/lib/python2.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
    """)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
===================== 1 warnings, 3 error in 58.18 seconds =====================
Traceback (most recent call last):
  File "./runner", line 55, in <module>
    subprocess.check_call(cmd, shell=True)
  File "/usr/lib64/python2.7/subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'docker run --net=host --privileged --volume=/data0/dev/ClickHouse/build/dbms/programs/clickhouse:/clickhouse         --volume=/usr/local/etc/clickhouse-server/:/clickhouse-config --volume=/data0/dev/ClickHouse:/ClickHouse -e PYTEST_OPTS='test_replace_partition -ss' yandex/clickhouse-integration-tests-runner' returned non-zero exit status 1
```

**Comment by nicelulu** [comment]
It looks like it is missing libreadline.so.6.

**Pull request merged**